flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -427,6 +427,28 @@ class JobManager(
           )
       }
 
+    case StopJob(jobID) =>
+      log.info(s"Trying to stop job with ID $jobID.")
+
+      currentJobs.get(jobID) match {
+        case Some((executionGraph, _)) =>
+          try {
+            if (executionGraph.getState() != JobStatus.RUNNING) {
```

Comment:
This guard is not sufficient since job state changes can happen asynchronously. Thus, it might happen that the current state is `JobStatus.RUNNING` when checking this condition, but before you execute `executionGraph.stop()` it will change to `JobStatus.RESTARTING`, for example. As a consequence, the sender will receive a `StoppingSuccess` message even though the job won't be stopped.